### ISSUE TITLE
Add a StringKeyedDictionary<T> type

### ISSUE THREAD

**Issue by TylerBrinkley** [dotnet/corefx#31942]
Moved from [corefxlab#2438](https://github.com/dotnet/corefxlab/issues/2438).

## Motivation
I would like a `StringKeyedDictionary<T>` type added in order to allow using a `ReadOnlySpan<char>` value for dictionary lookups without having to allocate a `string`.

## Proposed API
```diff
 namespace System.Collections.Specialized {
+    public class StringKeyedDictionary<T> : Dictionary<string, T> {
+        public StringKeyedDictionary();
+        public StringKeyedDictionary(IStringEqualityComparer comparer);
+        public StringKeyedDictionary(int capacity);
+        public StringKeyedDictionary(int capacity, IStringEqualityComparer comparer);
+        public StringKeyedDictionary(IDictionary<string, T> dictionary);
+        public StringKeyedDictionary(IDictionary<string, T> dictionary, IStringEqualityComparer comparer);
+        public StringKeyedDictionary(IEnumerable<KeyValuePair<string, T>> collection);
+        public StringKeyedDictionary(IEnumerable<KeyValuePair<string, T>> collection, IStringEqualityComparer comparer);
+        public new IStringEqualityComparer Comparer { get; }
+        public T this[ReadOnlySpan<char> key] { get; }
+        public bool ContainsKey(ReadOnlySpan<char> key);
+        public bool Remove(ReadOnlySpan<char> key);
+        public bool TryGetValue(ReadOnlySpan<char> key, out T value);
+    }
+    public interface IStringEqualityComparer : IEqualityComparer<string> {
+        bool Equals(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
+        int GetHashCode(ReadOnlySpan<char> obj);
+    }
 }
 namespace System {
-     public abstract class StringComparer : IComparer<string>, IEqualityComparer<string>, IComparer, IEqualityComparer {
+     public abstract class StringComparer : IComparer<string>, IEqualityComparer<string>, IComparer, IEqualityComparer, IStringEqualityComparer {
+        public virtual bool Equals(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
+        public virtual int GetHashCode(ReadOnlySpan<char> obj);
      }
     public sealed class CultureAwareComparer : StringComparer {
+        public override bool Equals(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
+        public override int GetHashCode(ReadOnlySpan<char> obj);
     }
     public class OrdinalComparer : StringComparer {
+        public override bool Equals(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
+        public override int GetHashCode(ReadOnlySpan<char> obj);
     }
 }
```

## Implementation Details
* The default implementation of the new virtual methods on `StringComparer` would convert the span to a string and use the string methods instead.
* Internal details of `Dictionary<TKey, TValue>` would be needed to implement the `ReadOnlySpan<char>` overloads.
* A `null` comparer passed to the constructor would default to `StringComparer.Ordinal`.
* Would need to be implemented in [corefx](https://github.com/dotnet/corefx) to access internal `Dictionary<TKey, TValue>` details as well as update `StringComparer`.

## Possible Addition
While not specifically needed for this request an `IStringComparer` interface should probably be added as well like so since we're updating `StringComparer`. It could then be used in a specialized `StringKeyedSortedDictionary` or `StringSortedSet`.

```diff
 namespace System.Collections.Specialized {
+    public interface IStringComparer : IComparer<string> {
+        int Compare(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
+    }
 }
 namespace System {
-    public abstract class StringComparer : IComparer, IEqualityComparer, IComparer<string>, IEqualityComparer<string>, IStringEqualityComparer {
+    public abstract class StringComparer : IComparer, IEqualityComparer, IComparer<string>, IEqualityComparer<string>, IStringEqualityComparer, IStringComparer {
+        public virtual int Compare(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
     }
     public sealed class CultureAwareComparer : StringComparer {
+        public override int Compare(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
     }
     public class OrdinalComparer : StringComparer {
+        public override int Compare(ReadOnlySpan<char> x, ReadOnlySpan<char> y);
     }
 }
```

## Open Questions
* Should the additional `IStringComparer` interface also be included?

**Comment by svick** [comment]
Would it be better to instead have a set of extension methods on `Dictionary<string, T>`? That way, the new methods could be used on any `string`-key `Dictionary`, not just on some special type.

**Comment by Wraith2** [comment]
`Dictionary<TKey,TValue>` is implemented in Coreclr and the is a high bar for new types because of library size issues. Do you intend or this implementation not to be collision resistant or do you have a method to store the key without the span somehow? convert it into a ReadOnlyMemory?

**Comment by benaadams** [comment]
> Would it be better to instead have a set of extension methods on Dictionary<string, T>? 

How would it perform the search (using hashcode + equals) against the current Dictionary api surface?

>  do you have a method to store the key without the span somehow? convert it into a ReadOnlyMemory?

`ROS<char>.ToString()`

**Comment by svick** [comment]
@benaadams 

>>    Would it be better to instead have a set of extension methods on Dictionary<string, T>?
>
> How would it perform the search (using hashcode + equals) against the current Dictionary api surface?

It wouldn't use the public API surface, it would use some new `internal` members on `Dictionary`. That's worse than `StringKeyedDictionary`, which would use new `protected` members, but I think not by much.

**Comment by benaadams** [comment]
Ah.. so something like?
```csharp
static class StringKeyedDictionaryExtensions 
{
    static bool ContainsKey<TValue>(ReadOnlySpan<char> key);
    static bool Remove<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key);
    static bool TryGetValue<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key, out TValue value);
}
```

**Comment by Wraith2** [comment]
Right, I think I missed the point then. The idea is to have the standard dictionary at the back end but methods which allow you to use ROS<char> as the key to search for avoiding the allocation to check if something exists. So it needs access to the internals because the comparer needs to understand that it can do ROS-string comparisons which aren't currently possible.

**Comment by svick** [comment]
@benaadams Yup. Plus probably an extension method version of the indexer:

```c#
public static T GetValue<TValue>(this Dictionary<string, TValue> dictionary, ReadOnlySpan<char> key);
```

Though that's an advantage of the `StringKeyedDictionary` design: it can have a real indexer.

**Comment by safern** [comment]
Also if they're extension methods we would need to provide overrides to specify a custom Comparer that can compare ROS right? Or that would be achieved by adding the StringEqualityComparer type and interface and in the extension methods check if Comparer is IStringEqualityComparer then use the ROS overrides if not the string ones? 

Also with override methods Adding values through the indexer or TryAdd/Add apis, we wouldn't be able to take advantage of the ROS comparisons right?

**Comment by benaadams** [comment]
> Also if they're extension methods we would need to provide overrides to specify a custom Comparer that can compare ROS right? 

Dictionary is based on `IEqualityComparer<T>` so not sure if it could work... (even if default interfaces implementations pulling up the slack, not sure how it could work for other types as a ROS comparison would be exposed on the interface, which probably wouldn't make sense?)

> Also with override methods Adding values through the indexer or TryAdd/Add apis, we wouldn't be able to take advantage of the ROS comparisons right?

Can covert the `string` to a ROS at nominal cost for the comparison; if it didn't exist the `ROS<char>` could be turned into a `string` for persistent key use; so that wouldn't necessarily be a problem.

**Comment by safern** [comment]
> Dictionary is based on IEqualityComparer<T> so not sure if it could work... (even if default interfaces implementations pulling up the slack, not sure how it could work for other types as a ROS comparison would be exposed on the interface, which probably wouldn't make sense?)

Yes, but people could just use `StringEqualityComparer` and since it would implement `IEqualityComparer` it can be passed through the constructor. However in order to use the ROS<char> overloads if we go through the Extension methods route we would need to cast the Comparer to `IStringEqualityComparer` 

I think if we want to use the ROS<char> comparison features and other of its features through a custom Comparer we should not use Extension methods, unless I’m missing something we could do in a different way without overriding implementation.

**Comment by TylerBrinkley** [comment]
Here's a comparison of the pro's and con's of the different design approaches.

|      | Extension Methods | StringKeyedDictionary |
| --- | --------------------- | ------------------------ |
| Pros | No new type introduced. | No performance penalty due to casting the comparer. |
| | Existing code exposing a string keyed `Dictionary` needs no changes to use which is helpful when they can't easily be changed. | Indexer support. |
| | | |
| Cons | Performance penalty of casting the comparer. | New type introduced meaning an additional staple of the .NET ecosystem to know. |
| | No indexer support, must use a `GetValue` method instead. | Existing code that exposes a string keyed `Dictionary` may not easily be changed to using the new type. |
| | `Dictionary` internals exposed with the `internal` modifier. | `Dictionary` internals exposed with the `private protected` modifier. |
| | Somewhat opaque to user as to knowing it uses `IStringEqualityComparer`. | |

Honestly, I'm kind of torn on this. Adding a new foundational Type to the ecosystem is costly in terms of what developers need to know. I'm feeling this even more now with the addition of all the types introduced with `Spans` and `Pipelines`. Additionally, I could easily see this expanding to others such as `Hashset`, `SortedDictionary`, `SortedSet`, `OrderedDictionary` & `OrderedSet` if they ever get implemented, and `ConcurrentDictionary`.

If the performance isn't an issue I now think I'd prefer the extension method route.

**Comment by safern** [comment]
@TylerBrinkley thanks for the great summary, I just updated the issue description and included this comparison in there. Will mark as ready for review now.

**Comment by cdmihai** [comment]
Overloads for `ConcurrentDictionary` would also be nice :)

**Comment by terrajobst** [comment]
On the surface, the extension approach looks more appealing because it would allow the lookups to work against any instance of a dictionary (at least in principle, the current design only works if the dictionary was instantiated with a comparer that implements `IStringEqualityComparer`).

However, this also complicates the implementation and might have performance implications if we have to expose more methods that only make sense when `TKey` happens to be `string`. Also, one could argue that in the hot paths where people would be inclined to lookup with spans that they do in fact control their dictionary and whether they need to pass in the correct comparer or use a derived type of `Dictionary<,>` doesn't matter.

I totally buy the scenario, but adding a new collection type (even if derived from `Dictionary<,>`) feels heavy handed, but maybe I'm overthinking this. Maybe just putting it in `S.C.Specialized` is good enough.

What do others think?

**Comment by jkotas** [comment]
This is trying to workaround the fundamental problem that ref-like types cannot be used as keys in collections today. We will hit the same problem for all keyed collections and many ref-like types. @cdmihai mentioned that `ConcurrentDictionary` would be nice too. The same can be said for `HashMap` and other collections. And the same can be said for Utf8 strings, Spans and other ref-like types. If we take this proposal, where are we going to stop – are we going to add all combinations like StringKeyedHashMap, SpanKeyedConcurrentDictionary?

I think we need to have [ref struct](https://github.com/dotnet/csharplang/issues/1148) constraint first to make this scalable.

**Comment by TylerBrinkley** [comment]
I agree scalability is an issue if we add new types which is why I'm now in favor of the extension methods route which should scale more favorably.

Even if the ref struct constraint is added I don't see how we can use that for this use case in a non-stack context such as class members or static fields.

**Comment by Joe4evr** [comment]
> This is trying to workaround the fundamental problem that ref-like types cannot be used as keys in collections today.

It's not so much using `ReadOnlySpan<char>` as the dictionary key, but doing a lookup with it without copy-allocating the slice it is wrapping.

**Comment by benaadams** [comment]
Is what's equatable; so `string` <=> `ReadOnlySpan<char>`; however what about `string` <=> `ReadOnlySpan<byte>, Encoding.UTF8` etc

**Comment by raffree** [comment]
I would change NameValueCollection, but more modern version, passing allowDublicates parameter in constructor. 
For dictionary i think extension method will fit completely.

**Comment by TylerBrinkley** [comment]
> what about `string` <=> `ReadOnlySpan<byte>, Encoding.UTF8` etc

Another reason the extension methods route makes better sense.

What if we instead relied on the comparer to derive from `StringComparer` and not define an interface? That way we can easily expand support through adding virtual methods without having to define new interfaces or else wait for default interface implementation support.

**Comment by Joe4evr** [comment]
> what about `string` <=> `ReadOnlySpan<byte>, Encoding.UTF8` etc

> What if we instead relied on the comparer to derive from `StringComparer` and not define an interface?

Sounds reasonable. I can already see use-cases for wanting to determine the value equality between a `string` and `utf8string`.
```cs
//using regular old UTF-16 strings as a key due to framework reasons,
//even though 99.9% of the time, it doesn't go outside UTF-8 range
private readonly Dictionary<string, Foo> _fooMap = new Dictionary<string, Foo>(StringComparer.OrdinalIgnoreCase);

//what this proposal already boils down to:
void M(ReadOnlySpan<char> sliceOfString)
{
    if (_fooMap.TryGetValue(sliceOfString, out var foo)
    {
        //....
    }
}

//but when incoming data could also be UTF-8:
void M(ReadOnlySpan<byte> sliceOfUtf8String) //or 'ROS<Char8>' if that becomes the more appropriate thing
{
    if (_fooMap.TryGetValue(sliceOfUtf8String, out var foo)
    {
        //....
    }
}
```

**Comment by TylerBrinkley** [comment]
Really appreciated hearing the discussion of this in the (.NET Design Reviews GitHub Triage)[https://youtu.be/hLbCfJSPIaI?t=5534].

I think this should go back to triage to decide the best design approach. Again, personally I'd much prefer the extension method route.

**Comment by DaZombieKiller** [comment]
What if an API was added to Dictionary<TKey, TValue> that allows you to retrieve a bucket?
For example (where DictionaryBucket is a new type, mainly to ensure that it's read-only):

`DictionaryBucket<TValue> Dictionary<TKey, TValue>.GetBucket(int hashCode)` or
`IReadOnlyList<TValue> Dictionary<TKey, TValue>.GetBucket(int hashCode)`

This way, the extension methods could operate on public APIs, and would just need to iterate over the bucket and compare values.

I'm not sure how realistic this would be, and I'm absolutely certain that there are significant problems with this that I haven't considered, but I figured I'd throw the idea out there.

**Comment by Wraith2** [comment]
It would be exposing and locking implementation details of the dictionary, it would prevent reimplementing using another method if a better one were found in future.

**Comment by TylerBrinkley** [comment]
As `Dictionary` doesn't currently support this scenario I was looking at implementing this in a custom dictionary and am having trouble getting the hashcode of the span. Are there any API's currently exposed which allows the retrieval of the hashcode of a `ReadOnlySpan<char>` in the `Ordinal` or `OrdinalIgnoreCase` sense without converting to a string?

**Comment by jkotas** [comment]
Look at the APIs introduced in https://github.com/dotnet/corefx/issues/31302

**Comment by TylerBrinkley** [comment]
@jkotas Thanks. Since .NET Standard 2.1 is including some API's introduced for .NET Core 3.0 would including this API be considered as well?

**Comment by jkotas** [comment]
.NET Standard 2.1 is very close to done at this point. We do not plan to add more APIs to it.

APIs introduced in .NET Core 3.0 are not included in .NET Standard 2.1 by default. Exceptions were made for a few APIs, mostly the ones required to support C# 8.0.

**Comment by TylerBrinkley** [comment]
@jkotas Thanks. I guess I'll resort to multi-targeting .NET Core App as well then.

**Comment by scalablecory** [comment]
I would prefer a generalized "proxy key lookup" support on `Dictionary`/etc. than specializing it for strings.

**Comment by TylerBrinkley** [comment]
@scalablecory what use case would you have for a generalized "proxy key lookup"? Also, what kind of API would you propose that would address this use case given the nature of the stack-only `ReadOnlySpan<char>`?

**Comment by scalablecory** [comment]
@TylerBrinkley we will need to combine a `ref struct` generic constraint with a new proposal for [heterogeneous lookup](https://www.bfilipek.com/2019/05/heterogeneous-lookup-cpp14.html) on keyed containers.

Special-casing some support for `String` would be a band-aid that we can't remove later on, and I'd rather wait for language changes needed to make a proper generalized change.

**Comment by TylerBrinkley** [comment]
@scalablecory Are you thinking of something like this?

```c#
public interface IAltEqualityComparer<T, TAlt> : IEqualityComparer<T> {
    public int GetHashCode(TAlt value);
    public bool Equals(TAlt value, T other);
}
public interface IAltRefEqualityComparer<T, TAlt> : IEqualityComparer<T> where TAlt : ref struct {
    public int GetHashCode(TAlt value);
    public bool Equals(TAlt value, T other);
}
public class Dictionary<TKey, TValue> {
    public bool ContainsKey<TAltKey>(TAltKey altKey);
    public bool ContainsKeyRef<TAltKey>(TAltKey altKey) where TAltKey : ref struct;
}
```

or this?

```c#
public delegate int GetHashCodeDelegate<TAltKey, TKey>(TAltKey altKey, IEqualityComparer<TKey> comparer);
public delegate bool EqualsDelegate<TAltKey, TKey>(TAltKey altKey, TKey key, IEqualityComparer<TKey> comparer);
public delegate int GetHashCodeDelegateRef<TAltKey, TKey>(TAltKey altKey, IEqualityComparer<TKey> comparer) where TAltKey : ref struct;
public delegate bool EqualsDelegateRef<TAltKey, TKey>(TAltKey altKey, TKey key, IEqualityComparer<TKey> comparer) where TAltKey : ref struct;

public class Dictionary<TKey, TValue> {
    public bool ContainsKey<TAltKey>(TAltKey altKey, GetHashCodeDelegate<TAltKey, TKey> getHashCode, EqualsDelegate<TAltKey, TKey> equals);
    public bool ContainsKey<TAltKey>(TAltKey altKey, GetHashCodeDelegateRef<TAltKey, TKey> getHashCode, EqualsDelegateRef<TAltKey, TKey> equals) where TAltKey : ref struct;
}
```

**Comment by scalablecory** [comment]
@TylerBrinkley I'm imagining a hybrid of the two.

### C# changes - `ref struct` constraint

Need a better name for this, and a C# proposal. Maybe a CLR proposal too...

The constraint would enforce that values of the type are only used in ways which are compatible with `ref struct` types.

So, you can pass in either a normal type or a `ref struct`, but it can't be stored in a non-stack field, can't be captured or used in async, etc.

### Library changes

```c#
public interface IEqualityComparer<T, TAlt> where TAlt: ref struct
{
    int GetHashCode(TAlt value);
    bool Equals(TAlt value, T other);
}

class StringComparer : IEqualityComparer<string, ReadOnlySpan<char>>
{
    public int GetHashCode(ReadOnlySpan<char> value);
    public bool Equals(ReadOnlySpan<char> value, string other);
}

class Dictionary<TKey, TValue>
{
    public bool ContainsKey<TAltKey>(TAltKey key, IEqualityComparer<TKey, TAltKey> comparer) where TAltKey: ref struct;
    public bool TryGetValue<TAltKey>(TAltKey key, IEqualityComparer<TKey, TAltKey> comparer, out TValue value) where TAltKey: ref struct;
    public bool Remove<TAltKey>(TAltKey key, IEqualityComparer<TKey, TAltKey> comparer) where TAltKey: ref struct;
    public bool Remove<TAltKey>(TAltKey key, IEqualityComparer<TKey, TAltKey> comparer, out TValue value) where TAltKey: ref struct;
}

class HashSet<T>
{
    public bool Contains<TAlt>(TAlt item, IEqualityComparer<T, TAlt> comparer) where TAlt: ref struct;
    public bool Remove<TAlt>(TAlt item, IEqualityComparer<T, TAlt> comparer) where TAlt: ref struct;
    public bool TryGetValue<TAlt>(TAlt item, IEqualityComparer<T, TAlt> comparer, out T actualValue) where TAlt: ref struct;
}
```

### Risks

Passing in an incompatible comparer might cause these to return an incorrect value. Usage of these will only happen in more advanced scenarios, so I think this is acceptable.

**Comment by TylerBrinkley** [comment]
I really don't like passing in the comparer in the method call due to the incompatible comparer issues and general usability. Is there a reason you don't suggest having the methods just try to type cast the collection's comparer?

**Comment by scalablecory** [comment]
> I really don't like passing in the comparer in the method call due to the incompatible comparer issues and general usability. Is there a reason you don't suggest having the methods just try to type cast the collection's comparer?

I agree, it's not good for usability and I don't like it. Mainly looking for feedback :).

I'd really like to find ways to avoid a cast, because:
a) adds a cast in what will be used in perf-sensitive code. (may or may not be significant depending on type/data -- need to measure)
b) moves usage errors from compile-time to run-time.

To look at an existing implementation, C++ has something in between these two approaches:
* Everything is baked into the comparer.
* A default comparer supporting this (`std::less<void>`) uses ADL and thus has the same usability concerns.
* ... but sidesteps those concerns because it resolves all of this at compile time rather than run time.

**Comment by MihaMarkic** [comment]
There is also `ImmutableDictionary<TKey, TValue>` out there and other types. Implementing a specialized dictionary type wouldn't solve all the others.

**Comment by DaZombieKiller** [comment]
Now that we have default interface implementations, would it make sense for this to be done by adding new APIs to `IDictionary`/`IReadOnlyDictionary` (among others) with default implementations that allocate and call the regular methods?
